### ISSUE TITLE
ngAnimate cause errors after updating to 1.3 (Cannot read property 'parentNode' of undefined)

### ISSUE THREAD

**Issue by hamfastgamgee** [angular/angular.js#10205]
I think the guy who filed #9953 gave up too easily. :)  I'm seeing the same error message in Angular 1.3.  I'm using a modified version of the angular-ui bootstrap datepicker that uses ng-if to only render the control if it is open, and so I'm reasonably certain the analysis by @caitp in that issue is exactly the problem for me.  If I add the following at line 1201 in angular-animate.js after the call to extractElementNode(), the error message goes away:

```javascript
if (!elementNode) {
    done();
    return;
}
```

There's also a comment by @matsko there about refactoring some code into animationsDisabled() after #8092 is fixed, which it now is, so I'm not sure if there's some more work to be done in that area.

I have not noticed any consequences to this, since presumably we can't do much to animate a DOM element that isn't there...

**Comment by hamfastgamgee** [comment]
Okay, I've managed to work around it in the datepicker in the same way (adding a <div> to the template).  I'm thinking this is probably another side effect of whatever that bug is regarding ngAnimate issues if you have replace: true and ng-if on the first line of the template.

**Comment by matsko** [comment]
Sorry for the delays everyone. I will be reviewing this issue tomorrow.

**Comment by crisbeto** [comment]
I got this error, because I had `ng-show` on a `ng-message` element:

```html
<div ng-messages="printWizardForm.title">
    <div class="form-error" ng-message="required" ng-show="form.$submitted">This field is required</div> <!-- error -->
    <div class="form-error" ng-message="maxlength">Input is too long</div> <!-- no error -->
</div>
```

**Comment by rochdev** [comment]
I am also getting this error with just an empty directive with `transclude: 'element'`.

**Comment by caitp** [comment]
this will happen with element transclusion, if you don't arrange things just right, because the "element" will be a comment node, which is ignored by ngAnimate

**Comment by rochdev** [comment]
So it is not possible to support `ngModel` on a directive with `transclude: element` when `ngAnimate` is loaded?

See http://plnkr.co/edit/vrl5ATDXNkE5v4LrFwDE

**Comment by staeke** [comment]
I've reproduced a(nother) case that looks like the same bug. ```ngClass``` never manages to set a class since parentNode is undefined and shortcuts ```runAnimationPostDigest```. This bug is currently exposed e.g. in angular-foundation. Requirements for the bug

- ngAnimate is imported and dependency declared (triggers the buggy logic)
- There is a "dynamic" directive that is compiled "off-DOM"
- This directive uses ```replace:true``` (means no normal parent node)
- This directive is linked with a ```cloneAttachFn``` (means no document fragment)
- It is not added to the DOM immediately (done with ```$timeout``` here, often the case for popups and similar)

Plunkr to reproduce http://plnkr.co/edit/rTfUUjkemyl7xfBwjn58?p=preview

**Comment by staeke** [comment]
Is that a good enough plunkr for you @matsko

**Comment by staeke** [comment]
Any idea of when you have time to look at this or if you think I should open another case for this @matsko ?

**Comment by matsko** [comment]
This is fixed for 1.4 (everything was refactored). Once that's merged in next week then I can fix it for 1.3.

**Comment by staeke** [comment]
ah great

**Comment by BenSmartcheckups** [comment]
What is the status on the fix for 1.3 ?

**Comment by crisbeto** [comment]
Was this fixed in 1.3.15? I just updated and it doesn't seem to be happening anymore.

**Comment by felipe-pereira** [comment]
I'm using 1.3.5, the issue is still present.

**Comment by stephengardner** [comment]
`ng-show` on an element with a directive whose directive returns transclusion.

In my case, I wrapped the directive `div` in another `div`, and applied the `ng-show` to that outer `div`.

Hopefully that helps someone else snake through their code a little easier.  Isolating the case by iteratively applying `ng-if="0"` to the elements in your views may help isolate it as well.

**Comment by ansgarkroger** [comment]
We are experiencing this issue with 1.3.17. Replacing 
[line 1999](https://github.com/angular/angular.js/blob/v1.3.x/src/ngAnimate/animate.js#L1199) `var parentNode = elementNode.parentNode;`
 with
`var parentNode = elementNode === undefined ? undefined : elementNode.parentNode;`
seems to work.

**Comment by crisbeto** [comment]
If I remember correctly, this was fixed in 1.4.0.

**Comment by e-cloud** [comment]
but there are a lot more that using 1.3.x

**Comment by gerardp** [comment]
I'm using 1.3.17 and i have the same error, please fix it !

**Comment by samy100** [comment]
Where's the fix????

**Comment by RamonBeast** [comment]
+1

**Comment by kichooo** [comment]
+1 ;)

**Comment by yeahoffline** [comment]
+1

**Comment by Blackbaud-PatrickOFriel** [comment]
I'd love to get a fix for 1.3.x as well

**Comment by matsko** [comment]
Here's a pending fix: https://github.com/angular/angular.js/pull/12603

**Comment by matsko** [comment]
Alright we have this merged into 1.3.x: https://github.com/angular/angular.js/commit/2c03a3574336ed814d020cf7ba36cee5b87e65b5

We should have a new release of 1.3 this week (hopefully tomorrow).

**Comment by matsko** [comment]
FYI folks, 1.3.18 has been released and it contains the fix: https://code.angularjs.org/1.3.18/

**Comment by Blackbaud-PatrickOFriel** [comment]
You're the man!

**Comment by mrasband** [comment]
This is awesome, we were having this issue yesterday and just upped our angular version. Can confirm that this fixed the issue as we were seeing it.

**Comment by elishaterada** [comment]
This fixed the issue on my project as well

**Comment by smarbos** [comment]
I was having the same problem.
i have a ng-hide in the same element that has a pagination directive.

**Comment by matsko** [comment]
@smarbos this was fixed with 1.3.18 yes?

**Comment by smarbos** [comment]
I know. Just saying.... In my case i can't update angular.So i fix it
putting the ng-show in another div.
Thank you!

**Comment by matsko** [comment]
You couldn't update at the time or you still can't update right now? 1.3.x releases don't have breaking changes so you should be fine. Or is there something more complicated?

**Comment by mariohmol** [comment]
I had this issue but i was using bower install --save ngAnimate instead of bower install --save angular-animate , fix it after using angular-animate

**Comment by noway** [comment]
Still present in 1.6.4 with ui-bootstrap 2.5.0

**Comment by kj1981** [comment]
Is there any working solution that works with Angular 1.3.4, NgAnimate 1.3.0, angular-ui-bootstrap/0.13.4
Sorry but at this moment I cannot update these libraries so wanted to know if there is a workaround/hack for this because I am struggling to find solution or alternative. Below is my codepen if it helps: https://codepen.io/anon/pen/wEaoLG
